guava/src/com/google/common/util/concurrent/MoreExecutors.java

```diff
@@ -841,4 +910,53 @@ static ScheduledExecutorService renamingDecorator(final ScheduledExecutorService
       }
     };
   }
+
+  /**
+   * Shuts down the given executor gradually, first disabling new submissions and later cancelling
+   * existing tasks.
+   *
+   * <p>The method takes the following steps:
+   * <ol>
+   *  <li>calls {@link ExecutorService#shutdown()}, disabling acceptance of new submitted tasks.
+   *  <li>waits for half of the specified timeout.
+   *  <li>if the timeout expires, it calls {@link ExecutorService#shutdownNow()}, cancelling
+   *  pending tasks and interrupting running tasks.
+   *  <li>waits for the other half of the specified timeout.
+   * </ol>
+   *
+   * <p>If, at any step of the process, the given executor is terminated or the calling thread is
+   * interrupted, the method calls {@link ExecutorService#shutdownNow()}, cancelling
+   * pending tasks and interrupting running tasks.
+   *
+   * @param service the {@code ExecutorService} to shut down
+   * @param timeout the maximum time to wait for the {@code ExecutorService} to terminate
+   * @param unit the time unit of the timeout argument
+   * @return {@code true} if the pool was terminated successfully, {@code false} if the
+   *     {@code ExecutorService} could not terminate <b>or</b> the thread running this method
+   *     is interrupted while waiting for the {@code ExecutorService} to terminate
+   * @since 17.0
+   */
+  @Beta
+  public static boolean shutdownAndAwaitTermination(
+      ExecutorService service, long timeout, TimeUnit unit) {
+    checkNotNull(unit);
+    // Disable new tasks from being submitted
+    service.shutdown();
+    try {
+      long halfTimeoutNanos = TimeUnit.NANOSECONDS.convert(timeout, unit) / 2;
+      // Wait for half the duration of the timeout for existing tasks to terminate
+      if (!service.awaitTermination(halfTimeoutNanos, TimeUnit.NANOSECONDS)) {
+        // Cancel currently executing tasks
+        service.shutdownNow();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Call `shutdown()` to reject any incoming tasks and `awaitTermination()` to wait for existing tasks to complete or until a timeout is reached. To stop the execution and processing of tasks, if necessary, call `shutdownNow()`. Check the return value of `awaitTermination()` and handle when the method times out while waiting for other threads to stop executing. See the following diffs on GitHub that make this code change: [diff1](https://github.com/glowroot/glowroot/commit/81de5a3699c59e0cf98d485c380c1034e6cd28a5#diff-c608d2c40790b0e0e5228e6647f69729caa756bbe9127169139406314b80897eL143), [diff2](https://github.com/apache/directory-kerby/commit/3ed0e7cd3b8a0c3ce29293153e4e72f422f0637f#diff-eb42f072849f520563ec2a686789a836195f18075cc222ef355c6ea710b0a360L84)

Learn more: [ExecutorService in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html)